### PR TITLE
Adds GOAWAY support to IIS

### PR DESCRIPTION
**Pull request by jkotalik**
Again part of our gRPC support for IIS.

### CODE REVIEW

File: src/Servers/IIS/AspNetCoreModuleV2/InProcessRequestHandler/managedexports.cpp [hunk 1]
```diff
@@ -601,4 +601,14 @@ http_reset_stream(
...
 }
 
+EXTERN_C __MIDL_DECLSPEC_DLLEXPORT
+HRESULT
+http_response_set_need_goaway(
+    _In_ IN_PROCESS_HANDLER* pInProcessHandler
+    )
+{
+    IHttpResponse4* pHttpResponse = (IHttpResponse4*)pInProcessHandler->QueryHttpContext()->GetResponse();
+    pHttpResponse->SetNeedGoAway();
```
Comment (Tratcher) [inline src/Servers/IIS/AspNetCoreModuleV2/InProcessRequestHandler/managedexports.cpp:611]: Need -> Needs?

File: src/Servers/IIS/IIS/test/Common.FunctionalTests/Inprocess/ResetTests.cs [hunk 2]
```diff
@@ -153,6 +153,31 @@ public async Task AppException_AfterHeaders_ResetInternalError()
...
+                .UseHttp2Cat(deploymentResult.ApplicationBaseUri, async h2Connection =>
+                {
+                    await h2Connection.InitializeConnectionAsync();
+
+                    h2Connection.Logger.LogInformation("Initialized http2 connection. Starting stream 1.");
+
+                    await h2Connection.StartStreamAsync(1, GetHeaders("/Goaway"), endStream: true);
+
+                    var frame = await h2Connection.ReceiveFrameAsync();
+                    Assert.Equal(Http2FrameType.GOAWAY, frame.Type);
```
Comment (Tratcher) [inline src/Servers/IIS/IIS/test/Common.FunctionalTests/Inprocess/ResetTests.cs:174]: There should be a VerifyGoAway method that also checks the last stream id.
https://github.com/dotnet/aspnetcore/blob/df5269fea23f310fd7730b9eb0da5cc910cc1e6c/src/Servers/HttpSys/test/FunctionalTests/Http2Tests.cs#L166

File: src/Servers/IIS/IIS/test/testassets/InProcessWebSite/Startup.cs [hunk 3]
```diff
@@ -1315,6 +1315,12 @@ public async Task Reset_DuringRequestBody_Resets(HttpContext httpContext)
             }
         }
 
+        public Task Goaway(HttpContext httpContext)
+        {
+            httpContext.Response.Headers["Connection"] = "close";
```
Comment (Tratcher) [inline src/Servers/IIS/IIS/test/testassets/InProcessWebSite/Startup.cs:1320]:
```suggestion
            httpContext.Response.Headers[HeaderNames.Connection] = "close";
```

File: src/Servers/IIS/AspNetCoreModuleV2/InProcessRequestHandler/managedexports.cpp (hunk 1 above)
Comment (jkotalik) [reply src/Servers/IIS/AspNetCoreModuleV2/InProcessRequestHandler/managedexports.cpp:611]: That's the IIS api name, I don't think it's worth changing at this point.

File: src/Servers/IIS/IIS/src/Core/IISHttpContext.cs [hunk 4]
```diff
@@ -385,6 +386,15 @@ public unsafe void SetResponseHeaders()
             // This copies data into the underlying buffer
             NativeMethods.HttpSetResponseStatusCode(_pInProcessHandler, (ushort)StatusCode, reasonPhrase);
 
+            if (HttpVersion >= System.Net.HttpVersion.Version20 && NativeMethods.HttpHasResponse4(_pInProcessHandler))
+            {
+                // Check if connection close is set, if so setting goaway
+                if (string.Equals(ConnectionClose, HttpResponseHeaders[HttpKnownHeaderNames.Connection], StringComparison.OrdinalIgnoreCase))
+                {
+                    NativeMethods.HttpSetNeedGoAway(_pInProcessHandler);
```
Comment (halter73) [inline src/Servers/IIS/IIS/src/Core/IISHttpContext.cs:394]: Why is this needed for gRPC? I'm pretty sure Kestrel won't send a GOAWAY if you set a "Connection: close" response header. What happens to concurrent streams?

Comment (Tratcher) [reply src/Servers/IIS/IIS/src/Core/IISHttpContext.cs:394]: GoAway isn't a gRPC ask, it was an ask from other partners trying to adopt HTTP/2. It was implemented by Http.Sys and IIS at the same time as the other HTTP/2 improvements.

It was first implemented this way in HttpSys and is now being ported to IIS.

Comment (davidfowl) [reply src/Servers/IIS/IIS/src/Core/IISHttpContext.cs:394]: Who asked? Can you add @halter73, @shirhatti and I to the thread?

Comment (halter73) [reply src/Servers/IIS/IIS/src/Core/IISHttpContext.cs:394]: This should be an explicit feature. If this aborting concurrent streams, that's very different that setting "Connection: close" on an HTTP/1.1 response. If it's draining gracefully, we should probably have a way to specify a timeout.

Comment (Tratcher) [reply src/Servers/IIS/IIS/src/Core/IISHttpContext.cs:394]: The API was implemented by Windows a while ago, it's too late to add parameters.

This is a graceful GoAway, it does not abort active streams.

The customer explicitly requested the `Connection: close` behavior for consistency with HTTP/1.1.

When we get around to this in Kestrel I would expect there to be a richer feature to let you specify the status and even message, but it might still have a simple fallback for `Connection: close`.

Comment (davidfowl) [reply src/Servers/IIS/IIS/src/Core/IISHttpContext.cs:394]: Even if it's implemented that way in Http.sys, why wouldn't we do it properly in ASP.NET Core and Connection: close like an implementation detail

Comment (Tratcher) [reply src/Servers/IIS/IIS/src/Core/IISHttpContext.cs:394]: Let's walk all the way through it after standup.

**Comment by jkotalik** [comment]
Closing as we want to implement a consistent feature across all servers.
